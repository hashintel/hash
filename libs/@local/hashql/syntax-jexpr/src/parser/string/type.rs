use core::ops::Range;

use hashql_ast::node::{
    id::NodeId,
    r#type::{
        IntersectionType, StructField, StructType, TupleField, TupleType, Type, TypeKind, UnionType,
    },
};
use hashql_core::symbol::Ident;
use winnow::{
    ModalParser, ModalResult, Parser as _,
    ascii::multispace0,
    combinator::{
        alt, cut_err, dispatch, fail, opt, peek, preceded, repeat, separated, separated_pair,
        terminated,
    },
    error::{AddContext, ParserError, StrContext, StrContextValue},
    token::any,
};

use super::{combinator::ws, context::Input, ident::parse_ident};
use crate::parser::string::path::parse_path;

fn parse_type_infer<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>,
{
    let context = input.state;

    "_".with_span()
        .map(|(_, span)| Type {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            kind: TypeKind::Infer,
        })
        .parse_next(input)
}

fn parse_type_tuple_field<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<TupleField<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    parse_type
        .with_span()
        .map(|(r#type, span)| TupleField {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            r#type,
        })
        .parse_next(input)
}

fn parse_type_struct_field<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<StructField<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    separated_pair(
        parse_ident,
        ws(cut_err(":").context(StrContext::Expected(StrContextValue::CharLiteral(':')))),
        parse_type,
    )
    .with_span()
    .map(|((name, r#type), span)| StructField {
        id: NodeId::PLACEHOLDER,
        span: context.span(span),
        name,
        r#type,
    })
    .parse_next(input)
}

fn parse_type_paren_empty_tuple<'heap, 'span, 'source, E>(
    start_span: Range<usize>,
) -> impl ModalParser<Input<'heap, 'span, 'source>, Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    move |input: &mut Input<'heap, 'span, 'source>| {
        let context = input.state;

        cut_err(')')
            .span()
            .map(|span| {
                let span = context.span(Range {
                    start: start_span.start,
                    end: span.end,
                });

                Type {
                    id: NodeId::PLACEHOLDER,
                    span,
                    kind: TypeKind::Tuple(TupleType {
                        id: NodeId::PLACEHOLDER,
                        span,
                        fields: context.heap.vec(None),
                    }),
                }
            })
            .context(StrContext::Expected(StrContextValue::CharLiteral(')')))
            .parse_next(input)
    }
}

fn parse_type_paren_empty_struct<'heap, 'span, 'source, E>(
    start_span: Range<usize>,
) -> impl ModalParser<Input<'heap, 'span, 'source>, Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    move |input: &mut Input<'heap, 'span, 'source>| {
        let context = input.state;

        cut_err(":)")
            .span()
            .map(|span| {
                let span = context.span(Range {
                    start: start_span.start,
                    end: span.end,
                });

                Type {
                    id: NodeId::PLACEHOLDER,
                    span,
                    kind: TypeKind::Struct(StructType {
                        id: NodeId::PLACEHOLDER,
                        span,
                        fields: context.heap.vec(None),
                    }),
                }
            })
            .context(StrContext::Expected(StrContextValue::StringLiteral(":)")))
            .parse_next(input)
    }
}

fn parse_type_paren_struct<'heap, 'span, 'source, E>(
    ident: Ident<'heap>,
    partial_field_span: Range<usize>,

    start_span: Range<usize>,
) -> impl ModalParser<Input<'heap, 'span, 'source>, Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    move |input: &mut Input<'heap, 'span, 'source>| {
        let context = input.state;

        // to now construct the first field, we just need to parse the type
        let (field_type, field_range) = parse_type.with_span().parse_next(input)?;

        let field_span = context.span(Range {
            start: partial_field_span.start,
            end: field_range.end,
        });

        let mut field = Some(StructField {
            id: NodeId::PLACEHOLDER,
            span: field_span,
            name: ident,
            r#type: field_type,
        });

        let fields = terminated(
            repeat(0.., preceded(ws(","), parse_type_struct_field)),
            opt(","),
        );

        terminated(
            fields,
            ws(cut_err(')').context(StrContext::Expected(StrContextValue::CharLiteral(')')))),
        )
        .with_span()
        .map(move |(mut fields, mut span): (Vec<_>, _)| {
            let field = field.take().expect("Parser called more than once");

            span.start = start_span.start;
            let span = context.span(span);

            fields.insert(0, field);

            Type {
                id: NodeId::PLACEHOLDER,
                span,
                kind: TypeKind::Struct(StructType {
                    id: NodeId::PLACEHOLDER,
                    span,
                    fields: context.heap.transfer_vec(fields),
                }),
            }
        })
        .parse_next(input)
    }
}

fn parse_type_paren_tuple<'heap, 'span, 'source, E>(
    first: Type<'heap>,

    start_span: Range<usize>,
) -> impl ModalParser<Input<'heap, 'span, 'source>, Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let mut first = Some(first);

    move |input: &mut Input<'heap, 'span, 'source>| {
        let context = input.state;

        terminated(
            (
                cut_err(','),
                repeat(0.., terminated(parse_type_tuple_field, ws(","))),
                opt(parse_type_tuple_field),
            ),
            ws(cut_err(')').context(StrContext::Expected(StrContextValue::CharLiteral(')')))),
        )
        .with_span()
        .map(|((_, rest, last), mut span): ((_, Vec<_>, _), _)| {
            let first = first.take().expect("Parser called more than once");

            let first = TupleField {
                id: NodeId::PLACEHOLDER,
                span: first.span,
                r#type: first,
            };

            span.start = start_span.start;
            let span = context.span(span);

            let mut fields = Vec::with_capacity(rest.len() + 2);
            fields.push(first);
            fields.extend(rest);
            if let Some(last) = last {
                fields.push(last);
            }

            Type {
                id: NodeId::PLACEHOLDER,
                span,
                kind: TypeKind::Tuple(TupleType {
                    id: NodeId::PLACEHOLDER,
                    span,
                    fields: context.heap.transfer_vec(fields),
                }),
            }
        })
        .parse_next(input)
    }
}

fn parse_type_paren<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    enum ParseDecision<'heap> {
        Struct(Ident<'heap>, Range<usize>),
        TupleOrParen(Type<'heap>),
    }

    let context = input.state;

    let start_span = ws("(").span().parse_next(input)?;

    // when we're inside parenthesis we can be one of the following:
    // - a group
    // - a tuple
    // - a struct

    // take a look at the next token, to determine if what we're parsing an empty variant, in that
    // case we can return it
    match peek(any).parse_next(input)? {
        // empty struct
        ':' => return parse_type_paren_empty_struct(start_span).parse_next(input),
        // empty tuple
        ')' => return parse_type_paren_empty_tuple(start_span).parse_next(input),
        _ => {}
    }

    // Now that the empty variants are out of the way, we need to determine, are we in a struct,
    // tuple or paren?
    let decision = alt((
        terminated(parse_ident, ws(":"))
            .with_span()
            .map(|(ident, span)| ParseDecision::Struct(ident, span)),
        parse_type.map(ParseDecision::TupleOrParen),
    ))
    .parse_next(input)?;

    // given our parse decision we now know if we're in a struct, or not, in case we are, fall back
    // to the struct parser
    match decision {
        ParseDecision::Struct(ident, partial_field_span) => {
            parse_type_paren_struct(ident, partial_field_span, start_span).parse_next(input)
        }
        ParseDecision::TupleOrParen(r#type) => {
            // To finally figure out if we're in a paren or tuple, check the next character, if it's
            // a `)` we're in a paren, otherwise in a tuple

            let _: &str = multispace0.parse_next(input)?;

            match peek(opt(any)).parse_next(input)? {
                Some(')') => {
                    let mut r#type = Some(r#type);
                    cut_err(")")
                        .span()
                        .map(|mut span| {
                            let r#type = r#type.take().expect("Parser called more than once");

                            // Extend the span of the type to include the closing parenthesis
                            span.start = start_span.start;
                            context.cover(r#type.span, span);

                            r#type
                        })
                        .context(StrContext::Expected(StrContextValue::CharLiteral(')')))
                        .parse_next(input)
                }
                Some(',') => parse_type_paren_tuple(r#type, start_span).parse_next(input),
                _ => fail
                    .context(StrContext::Expected(StrContextValue::CharLiteral(')')))
                    .context(StrContext::Expected(StrContextValue::CharLiteral(',')))
                    .context(StrContext::Expected(StrContextValue::CharLiteral(':')))
                    .parse_next(input),
            }
        }
    }
}

fn parse_type_atom<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let mut path = parse_path.map(|path| Type {
        id: NodeId::PLACEHOLDER,
        span: path.span,
        kind: TypeKind::Path(path),
    });

    dispatch! {peek(any);
        '_' => parse_type_infer,
        '(' => parse_type_paren,
        _ => path
    }
    .parse_next(input)
}

fn parse_type_union<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    separated(1.., parse_type_atom, ws("|"))
        .with_span()
        .map(|(mut parsed, span): (Vec<_>, _)| {
            if parsed.len() == 1 {
                return parsed.pop().unwrap_or_else(|| unreachable!());
            }

            let span = context.span(span);

            let mut types = context.heap.vec(Some(parsed.len()));
            types.extend(parsed);

            Type {
                id: NodeId::PLACEHOLDER,
                span,
                kind: TypeKind::Union(UnionType {
                    id: NodeId::PLACEHOLDER,
                    span,
                    types,
                }),
            }
        })
        .parse_next(input)
}

fn parse_type_intersection<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    separated(1.., parse_type_union, ws("&"))
        .with_span()
        .map(|(mut parsed, span): (Vec<_>, _)| {
            if parsed.len() == 1 {
                return parsed.pop().unwrap_or_else(|| unreachable!());
            }

            let span = context.span(span);

            let mut types = context.heap.vec(Some(parsed.len()));
            types.extend(parsed);

            Type {
                id: NodeId::PLACEHOLDER,
                span,
                kind: TypeKind::Intersection(IntersectionType {
                    id: NodeId::PLACEHOLDER,
                    span,
                    types,
                }),
            }
        })
        .parse_next(input)
}

/// ```abnf
/// infer = "_"
/// tuple = "()" / "(" +(type ",") ?(type) ")"
/// struct = "(:)" / "(" ident ":" type *("," ident ":" type) ?"," ")"
/// paren = "(" type ")"
/// atom = path / tuple / struct / infer / paren
/// union = atom *("|" atom)
/// intersection = union *("&" union)
/// type = intersection
/// ```
pub(crate) fn parse_type<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Type<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    ws(parse_type_intersection)
        .context(StrContext::Label("type"))
        .parse_next(input)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::string::test::{bind_parser, test_cases};

    bind_parser!(SyntaxDump; fn parse_type_test(parse_type));

    // Tests for individual parser components
    test_cases!(parse_type_test;
        // Atom types
        infer_type("_") => "Infer type",
        simple_path("Int") => "Simple path type",
        qualified_path("Std::String") => "Qualified path",
        nested_qualified_path("Std::Collections::HashMap") => "Nested qualified path",

        // Tuple types
        empty_tuple("()") => "Empty tuple",
        single_element_tuple("(Int,)") => "Single-element tuple with trailing comma",
        multi_element_tuple("(Int, String, Boolean)") => "Multi-element tuple",

        // Struct types
        empty_struct("(:)") => "Empty struct",
        single_field_struct("(name: String)") => "Single-field struct",
        multi_field_struct("(name: String, age: Int)") => "Multi-field struct",

        // Union types
        simple_union("Int | String") => "Simple union type",
        multi_union("Int | String | Boolean") => "Multiple union type",

        // Intersection types
        simple_intersection("Int & String") => "Simple intersection type",
        multi_intersection("Int & String & Boolean") => "Multiple intersection type",
    );

    // Tests for composite structures and nesting
    test_cases!(parse_type_test;
        // Nested structures
        nested_tuple("((Int, String), Boolean)") => "Nested tuple type",
        nested_struct("(person: (name: String, age: Int))") => "Nested struct type",
        struct_with_tuple_field("(data: (Int, String))") => "Struct with tuple field",
        tuple_with_struct_field("((name: String), Int)") => "Tuple with struct field",

        // Whitespace handling
        tuple_with_whitespace("( Int , String )") => "Tuple with whitespace",
        struct_with_whitespace("( name : String , age : Int )") => "Struct with whitespace",
        union_with_whitespace(" Int | String ") => "Union with whitespace",
        intersection_with_whitespace(" Int & String ") => "Intersection with whitespace",
    );

    // Tests for operator precedence and complex expressions
    test_cases!(parse_type_test;
        union_of_atoms("Int | String | Boolean") => "Union of atom types",
        intersection_of_atoms("Int & String & Boolean") => "Intersection of atom types",
        union_of_intersections("Int & String | Boolean & Char") => "Union of intersections",
        intersection_of_unions("Int | String & Boolean | Char") => "Intersection of unions",
        mixed_precedence("Int & (String | Boolean) & Char") => "Mixed precedence with parentheses",
        deeply_nested("(((Int | String) & Boolean) | Char)") => "Deeply nested expression",
    );

    // Error cases
    test_cases!(parse_type_test;
        unclosed_tuple("(Int, String") => "Unclosed tuple",
        unclosed_struct("(name: String, age: Int") => "Unclosed struct",
        missing_field_type("(name: )") => "Missing field type",
        missing_colon("(name: String, age Int)") => "Missing field type",
        invalid_field_separator("(name String)") => "Invalid field separator",
        unclosed_parenthesized("(Int & String") => "Unclosed parenthesized type",
        incomplete_union("Int |") => "Incomplete union",
        incomplete_intersection("Int &") => "Incomplete intersection",
        invalid_operator("Int $ String") => "Invalid operator",
        double_operator("Int || String") => "Double union operator",
        missing_type("") => "Missing type",
    );

    // Tests for common real-world type patterns
    test_cases!(parse_type_test;
        generic_path("Container<T>") => "Generic type",
        generic_with_multiple_params("Map<Key, Value>") => "Generic with multiple parameters",
        nested_generics("Container<Wrapper<T>>") => "Nested generic types",
        optional_type("Option<Value>") => "Optional type pattern",
        complex_entity("(id: ID, attrs: (name: String, metadata: (created: Timestamp, modified: Timestamp | Null)))") => "Complex entity type with nested fields",
    );
}
