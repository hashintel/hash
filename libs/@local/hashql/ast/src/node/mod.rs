//! Core Abstract Syntax Tree node types for HashQL.
//!
//! This module defines the fundamental building blocks of the HashQL Abstract
//! Syntax Tree (AST). The AST represents the structure of a HashQL program after
//! parsing, capturing all the language constructs in a hierarchical tree structure.
//!
//! ## Module Organization
//!
//! The node module is organized into several submodules:
//!
//! - [`expr`]: Expression nodes representing computations that produce values
//! - [`generic`]: Generic type parameter and argument definitions
//! - [`id`]: Node identifier types and utilities
//! - [`path`]: Path expressions for variable references and qualified names
//! - [`type`]: Type system representations including struct, tuple, and other types
//!
//! ## Memory Management
//!
//! All AST nodes use arena allocation through the `'heap` lifetime parameter,
//! which references the memory region where nodes are allocated. This approach
//! improves performance by minimizing allocations and simplifies memory management
//! during parsing and analysis.
//!
//! ## Node Identification
//!
//! Each node in the AST has a unique [`id::NodeId`] that can be used to track
//! the node through various processing stages. Nodes also include a span identifier that points to
//! the source location of the node.
//!
//! ## AST Structure
//!
//! The AST is structured as a tree with expressions ([`expr::Expr`]) at the core.
//! Other node types like types ([`type::Type`]) and paths ([`path::Path`]) are
//! used as components within expressions or on their own for declarations.
//!
//! The entire tree represents a complete HashQL program or query, ready for
//! analysis, transformation, and evaluation.
pub mod expr;
pub mod generic;
pub mod id;
pub mod path;
pub mod r#type;
