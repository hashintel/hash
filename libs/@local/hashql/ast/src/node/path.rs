use hashql_core::{span::SpanId, symbol::Ident};

use super::{generic::GenericArgument, id::NodeId};
use crate::heap;

/// A segment in a path expression.
///
/// Represents a single component of a qualified path, consisting of an identifier
/// and optional generic type arguments. Path segments are connected to form
/// complete paths like `module::submodule::item<T>`.
///
/// # Examples
///
/// In a path like `std::collections::HashMap<K, V>`:
/// - `std` is a segment with no generic arguments
/// - `collections` is a segment with no generic arguments
/// - `HashMap` is a segment with generic arguments `K` and `V`
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct PathSegment<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    pub name: Ident,
    /// Type parameters attached to this path segment
    pub arguments: heap::Vec<'heap, GenericArgument<'heap>>,
}

/// A path expression in the HashQL Abstract Syntax Tree.
///
/// Represents a qualified reference to a variable, function, type, or module.
/// Paths can be simple identifiers (like `x`) or qualified names with multiple
/// segments separated by double colons (like `std::collections::HashMap`).
///
/// Path expressions are used for variable references, module access, and type names.
/// They can include generic type arguments applied to specific segments.
///
/// # Examples
///
/// Simple paths:
/// ```text
/// x                     // A local variable
/// length                // A function or variable
/// ```
///
/// Qualified paths:
/// ```text
/// std::collections::HashMap  // An item in a nested module
/// math::sin                  // A function in a module
/// ```
///
/// Paths with generic arguments:
/// ```text
/// Vec<Int>                       // A type with one type argument
/// HashMap<String, User>          // A type with two type arguments
/// graph::query<User, Connection> // A generic function
/// ```
///
/// J-Expr explicit:
/// ```json
/// {"#path", ["std", "collections", "HashMap"]}
/// {"#path", [{"name": "HashMap", "arguments": ["K, V"]}]}
/// ```
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Path<'heap> {
    pub id: NodeId,
    pub span: SpanId,

    /// Whether the path is rooted (starts with a double colon `::`).
    pub rooted: bool,
    pub segments: heap::Vec<'heap, PathSegment<'heap>>,
}
