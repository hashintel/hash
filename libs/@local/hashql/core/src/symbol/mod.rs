//! Symbol representation and manipulation for HashQL.
//!
//! This module defines dedicated types for representing symbols in HashQL's compilation process.
//! Symbols are string-like values that appear in source code, such as identifiers, keywords,
//! and literals. They are deliberately kept separate from the syntax tree as they are used through
//! the different stages of compilation.
//!
//! The module provides:
//!
//! - [`Symbol`]: An opaque wrapper around string data that enables efficient storage and comparison
//! - [`Ident`]: A named identifier with source location and categorization
//! - [`IdentKind`]: Classification of different identifier types in HashQL
//!
//! ## Design Philosophy
//!
//! The [`Symbol`] type is designed as an opaque wrapper around its internal string storage.
//! This encapsulation enables future optimizations such as string interning (either through
//! the `string_interner` crate or a custom implementation) without requiring API changes.

pub mod sym;

use core::{
    borrow::Borrow,
    cmp::Ordering,
    fmt::{self, Display, Formatter},
    hash::{Hash, Hasher},
    ptr,
};

use crate::span::SpanId;

/// A string-like value used throughout the HashQL compiler.
///
/// Symbols represent string data that appears in source code and persists throughout
/// compilation, they are read-only and immutable.
///
/// This type is deliberately opaque to hide its internal representation,
/// allowing for future optimizations like string interning without changing
/// the public API. Symbols are designed to be efficient for long-lived objects
/// that are frequently compared, hashed, and referenced during compilation.
///
/// The caller must ensure that the string is unique and interned. The types correctness requires
/// relies on these *but it does not enforce it*.
#[derive(Debug, Copy, Clone)]
pub struct Symbol<'heap>(&'heap str);

impl<'heap> Symbol<'heap> {
    /// Creates a new interned symbol from a string slice.
    ///
    /// The caller must ensure that the string is unique and interned.
    pub(crate) const fn new_unchecked(string: &'heap str) -> Self {
        Self(string)
    }

    #[must_use]
    pub const fn as_str(&self) -> &str {
        self.0
    }

    /// Returns the string representation of the symbol.
    ///
    /// Unlike [`Self::as_str`], this method provides access for the lifetime of the interner
    /// instead of the symbol itself, somewhat circumventing the protections given to the symbol
    /// itself. Any unwrapped type should be considered no longer unique and interned.
    #[must_use]
    pub const fn unwrap(&self) -> &'heap str {
        self.0
    }

    #[must_use]
    pub const fn as_bytes(&self) -> &[u8] {
        self.0.as_bytes()
    }
}

impl AsRef<Self> for Symbol<'_> {
    #[inline]
    fn as_ref(&self) -> &Self {
        self
    }
}

impl PartialEq for Symbol<'_> {
    fn eq(&self, other: &Self) -> bool {
        // Pointer equality implies string equality (due to the unique contents assumption)
        ptr::eq(self.0, other.0)
    }
}

impl Eq for Symbol<'_> {}

impl PartialOrd for Symbol<'_> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for Symbol<'_> {
    fn cmp(&self, other: &Self) -> Ordering {
        // Pointer equality implies string equality (due to the unique contents assumption), but if
        // not the same the contents must be compared.
        if self == other {
            Ordering::Equal
        } else {
            self.0.cmp(other.0)
        }
    }
}

impl Hash for Symbol<'_> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        // Pointer hashing is sufficient (due to the unique contents assumption)
        ptr::hash(self.0, state);
    }
}

impl Borrow<str> for Symbol<'_> {
    fn borrow(&self) -> &str {
        self.0
    }
}

impl AsRef<str> for Symbol<'_> {
    fn as_ref(&self) -> &str {
        self.0
    }
}

impl Display for Symbol<'_> {
    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
        Display::fmt(self.0, fmt)
    }
}

/// The classification of an identifier in HashQL.
///
/// HashQL supports different categories of identifiers, each with distinct
/// syntactic rules and semantic meanings in the language.
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum IdentKind {
    /// A lexical identifier following standard programming language naming rules.
    ///
    /// # Syntax
    ///
    /// Lexical identifiers in HashQL follow Rust's identifier parsing rules:
    /// - Begin with either an `XID_Start` Unicode character or an underscore (`_`)
    /// - Continue with zero or more `XID_Continue` Unicode characters
    /// - Cannot be a single underscore character (`_`)
    /// - Uppercase identifiers are used for types, while lowercase identifiers are used for values,
    ///   identifiers starting with an underscore (`_`) are considered lowercase.
    ///
    /// In formal notation:
    /// ```text
    /// IDENTIFIER_LEXICAL:
    ///     XID_Start XID_Continue*
    ///     | '_' XID_Continue+
    /// ```
    ///
    /// > Note: Zero-width characters are not allowed.
    ///
    /// > Note: All identifiers are normalized using Unicode Normalization Form C (NFC).
    ///
    /// # Examples
    ///
    /// ```text
    /// counter
    /// _identifier
    /// Москва        // Non-ASCII identifiers are supported
    /// 東京           // As are other scripts
    /// ```
    Lexical,

    /// A symbolic identifier consisting of operator-like characters.
    ///
    /// # Syntax
    ///
    /// Symbol identifiers can consist of ASCII operators or Unicode symbols/punctuation:
    ///
    /// ```text
    /// ASCII_SYMBOL:
    ///       '!' | '#' | '$' | '%' | '&'
    ///     | '*' | '+' | '.' | '/' | '<'
    ///     | '=' | '>' | '?' | '@' | '\'
    ///     | '^' | '|' | '-' | '~'
    ///
    /// RESERVED_SYMBOL:
    ///     ':'
    ///
    /// ; Excludes any symbols that are in the ASCII range
    /// UNICODE_SYMBOL:
    ///     Symbol
    ///     | Punctuation
    ///
    /// SYMBOL:
    ///     ASCII_SYMBOL
    ///     | UNICODE_SYMBOL
    ///
    /// IDENTIFIER_SYMBOL:
    ///     SYMBOL+
    ///     | '`' SYMBOL+ '`'
    /// ```
    ///
    /// > Note: `:` is reserved, and cannot be used as a symbol identifier.
    ///
    /// > Note: All identifiers are normalized using Unicode Normalization Form C (NFC).
    ///
    /// # Examples
    ///
    /// ```text
    /// <
    /// +
    /// ->
    /// ≥
    /// →
    /// `*`
    /// ```
    Symbol,

    /// An identifier representing a URL for external resource references.
    ///
    /// # Syntax
    ///
    /// Base URL identifiers must be valid `http` or `https` URLs that end with a forward slash
    /// (`/`), and must be enclosed within backticks:
    ///
    /// ```text
    /// IDENTIFIER_BASE_URL:
    ///     '`' HTTP_URL_STRING_ENDING_WITH_SLASH '`'
    /// ```
    ///
    /// > Note: All identifiers are normalized using Unicode Normalization Form C (NFC).
    ///
    /// # Examples
    ///
    /// ```text
    /// `https://blockprotocol.org/@blockprotocol/types/property-type/address/`
    /// `https://example.com/resources/schema/`
    /// `http://localhost:8080/api/v1/`
    /// ```
    BaseUrl,
}

/// A named identifier in HashQL with source location information.
///
/// An `Ident` combines a [`Symbol`] (the textual content of the identifier)
/// with a [`SpanId`] (its location in the source code) and a [`IdentKind`]
/// (its syntactic classification). Identifiers are used throughout the compiler
/// to represent variables, functions, types, and other named entities.
///
/// # Examples
///
/// In HashQL source code:
///
/// ```text
/// let counter = 0;
///     ^^^^^^^
/// ```
///
/// This identifier would be represented as an `Ident` with:
/// - name: `Symbol("counter")`
/// - kind: `IdentKind::Lexical`
/// - span: `5..12`
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct Ident<'heap> {
    pub span: SpanId,

    pub value: Symbol<'heap>,
    pub kind: IdentKind,
}

#[cfg(test)]
impl<'heap> Ident<'heap> {
    pub(crate) const fn synthetic(value: Symbol<'heap>) -> Self {
        Self {
            span: SpanId::SYNTHETIC,
            value,
            kind: IdentKind::Lexical,
        }
    }
}

impl AsRef<str> for Ident<'_> {
    fn as_ref(&self) -> &str {
        self.value.as_str()
    }
}

impl Display for Ident<'_> {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        Display::fmt(&self.value.0, fmt)
    }
}
