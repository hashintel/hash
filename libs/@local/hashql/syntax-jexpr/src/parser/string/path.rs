use hashql_ast::node::{
    id::NodeId,
    path::{Path, PathSegment},
};
use winnow::{
    ModalResult, Parser as _,
    combinator::{cut_err, delimited, opt},
    error::{AddContext, ParserError, StrContext, StrContextValue},
};

use super::{
    combinator::{separated_boxed1, ws},
    context::Input,
    generic::parse_generic_argument,
    ident::parse_ident,
};

fn parse_path_segment<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<PathSegment<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    let arguments = opt(delimited(
        ws("<"),
        separated_boxed1(context.heap, parse_generic_argument, ws(",")),
        ws(cut_err(">").context(StrContext::Expected(StrContextValue::CharLiteral('>')))),
    ));

    (parse_ident, arguments)
        .with_span()
        .map(|((ident, arguments), span)| PathSegment {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            name: ident,
            arguments: arguments.unwrap_or_else(|| context.heap.empty_slice()),
        })
        .parse_next(input)
}

pub(crate) fn parse_path<'heap, 'span, 'source, E>(
    input: &mut Input<'heap, 'span, 'source>,
) -> ModalResult<Path<'heap>, E>
where
    E: ParserError<Input<'heap, 'span, 'source>>
        + AddContext<Input<'heap, 'span, 'source>, StrContext>,
{
    let context = input.state;

    let root = opt(ws("::")).map(|value| value.is_some());
    let segments = separated_boxed1(context.heap, parse_path_segment, ws("::"));

    (root, segments)
        .with_span()
        .map(|((rooted, segments), span)| Path {
            id: NodeId::PLACEHOLDER,
            span: context.span(span),
            rooted,
            segments,
        })
        .context(StrContext::Label("path"))
        .parse_next(input)
}

#[cfg(test)]
mod tests {
    #![expect(clippy::non_ascii_literal)]
    use super::parse_path;
    use crate::parser::string::test::{bind_parser, test_cases};

    // Bind our parser to create a testing function
    bind_parser!(SyntaxDump; fn parse_path_test(parse_path));

    // Basic path structure tests
    test_cases!(parse_path_test;
        simple_identifier("foo") => "Simple identifier path",
        two_segment_path("foo::bar") => "Two-segment path",
        multi_segment_path("foo::bar::baz") => "Multi-segment path",
        rooted_path("::foo") => "Rooted path",
        rooted_multi_segment("::foo::bar") => "Rooted multi-segment path",
    );

    // Generic arguments
    test_cases!(parse_path_test;
        single_generic("Vec<T>") => "Single generic argument",
        multiple_generics("Map<K, V>") => "Multiple generic arguments",
        path_with_generics("std::collections::HashMap<K, V>") => "Path segments with generics",
        nested_generics("Result<Option<T>, E>") => "Nested generic arguments",
        complex_generics("HashMap<String, Vec<Option<T>>>") => "Complex nested generics",
    );

    // Whitespace variations
    test_cases!(parse_path_test;
        path_whitespace("foo :: bar") => "Whitespace around path separator",
        generics_whitespace("Vec< T , U >") => "Whitespace in generic arguments",
        mixed_whitespace("std :: collections :: HashMap< K , V >") => "Mixed whitespace patterns",
    );

    // Edge cases
    test_cases!(parse_path_test;
        unicode_identifier("标识符") => "Unicode identifier",
        mixed_unicode("std::标识符::Vec<T>") => "Mixed Unicode and ASCII in path",
        path_trailing_separator("foo::") => "Path with trailing separator",
        empty_root("::") => "Empty rooted path",
    );

    // Error handling
    test_cases!(parse_path_test;
        unclosed_generic("Vec<T") => "Unclosed generic argument",
        empty_generic_args("Vec<>") => "Empty generic arguments",
        missing_separator("Vec<T U>") => "Missing separator in generic arguments",
        invalid_separator("foo;bar") => "Invalid path separator",
        double_separator("foo::::bar") => "Double path separator",
        empty_path("") => "Empty path",
    );
}
