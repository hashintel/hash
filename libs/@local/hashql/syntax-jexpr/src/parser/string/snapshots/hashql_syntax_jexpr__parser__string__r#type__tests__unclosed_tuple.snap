---
source: libs/@local/hashql/syntax-jexpr/src/parser/string/type.rs
description: Unclosed tuple
expression: "(Int, String"
info:
  kind: Err
---
(Int, String
            ^
invalid type
expected `)`
